src/engine/controls/loopingcontrol.cpp

```diff
@@ -1192,15 +1211,26 @@ void LoopingControl::slotBeatLoop(double beats, bool keepStartPoint, bool enable
     } else {
         // loop_in is set to the closest beat if quantize is on and the loop size is >= 1 beat.
         // The closest beat might be ahead of play position and will cause a catching loop.
-        double prevBeat;
-        double nextBeat;
-        pBeats->findPrevNextBeats(currentSample, &prevBeat, &nextBeat, true);
+        mixxx::audio::FramePos prevBeatPosition;
+        mixxx::audio::FramePos nextBeatPosition;
+        const auto currentPosition = mixxx::audio::FramePos::fromEngineSamplePos(currentSample);
+        pBeats->findPrevNextBeats(currentPosition, &prevBeatPosition, &nextBeatPosition, true);
+        const double prevBeat = prevBeatPosition.isValid()
+                ? prevBeatPosition.toEngineSamplePos()
+                : -1;
+        const double nextBeat = nextBeatPosition.isValid()
```

Comment:
If `-1`is the corresponding engine sample value for invalid frame positions we might consider to return it from `toEngineSamplePos()`. I suppose this function will disappear eventually?

This conversion appears multiple times (see also: quantizecontrol.cpp), so we should at least wrap it into a utility function instead of repeating it over and over again.

Reply:
I agree. Unfortunately I only had the idea later.

Check out https://github.com/mixxxdj/mixxx/pull/4061/commits/6d13725c
314b1aca599409d770eb52d774fa4c69 and https://github.com/mixxxdj/mixxx/pull/4061/commits/b0be360f953189e7857ac4f599b3ca953409ba58. Both are part of https://github.com/mixxxdj/mixxx/pull/4061 which is a follow up to this PR. I'm not particularity keen on cherry-picking these commits here and causing merge conflicts. But I can do it if you want.